Pbc dist bench

Miro-Astore opened the pull request:
Fixes #3206 and #2794

Changes made in this Pull Request:
 - Added, for relevant functions, benchmarks for testing orthogona and triclinic 
 - PBC settings
 - Also removed 10000 atom benchmarks for contact matrix
functions to avoid timeout failiures.

Had to split the DistancesBench class into 3 classes to avoid
failiures and uncessary reruns of the benchmark. These failiures
were due to time outs when too many atoms were run in the
benchmarks or distances.between not taking box type as an argument.
The latter case meant, to avoid failiure, we had to either split
the classes out or run redundant benchmarks. I chose the latter
and created BetweenBench.

The 3 classes and the distribution of testing functions is

BetweenBench
   - time_between
DistancesBench
   - time_distance_array
   - time_distance_array_pre_allocated
   - time_self_distance_array
   - time_self_distance_array_pre_allocated
   - time_dist
   - time_dist_offsets
ContactsBench
   - time_contact_matrix
   - time_contact_matrix_sparse

Splitting the classes like this was necessary because asv does not
allow us to easily skip certain combinations of parameters at the
function level, only in the setup()
function of the class.


PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

Miro-Astore committed:
changed version of python specified by asv.conf.json to 3.7

Miro-Astore committed:
Added tests of different PBC conditions and fixed failing distances benchmarks

The benchmarked distances functions now all test 3 PBC conditions:
None, triclinic, orthogonal.

Separated matrix contacts functions and simple distance functions into
DistanceBench and ContactsBench classes to deal with time complexity issues.

Miro-Astore committed:
Coverage of different PBCs for distance benchmarks

Added, for relevant functions, benchmarks for testing orthogonal PBC
settings and also removed 10000 atom benchmarks for contact matrix
functions to avoid timeout failiures.

Had to split the DistancesBench class into 3 classes to avoid
failiures and uncessary reruns of the benchmark. These failiures
were due to time outs when too many atoms were run in the
benchmarks or distances.betweeen not taking box type as an argument.
The latter case meant, to avoid failiure, we had to either split
the classes out or run redundant benchmarks.

The 3 classes and the distribution of testing functions is

BetweenBench
    time_between
DistancesBench
    time_distance_array
    time_distance_array_pre_allocated
    time_self_distance_array
    time_self_distance_array_pre_allocated
    time_dist
    time_dist_offsets
ContactsBench
    time_contact_matrix
    time_contact_matrix_sparse

Splitting the classes like this was necessary because asv does not
allow us to easily skip certain combinations of parameters at the
function level, only in the setup()
function of the class.

Miro-Astore committed:
Fixing mistake in asv.conf.json where version was changed from 2.7 to 3.7, changed back to 2.7

Miro-Astore committed:
Fixing formatting of distances benchmark

pep8speaks commented:
Hello @Miro-Astore! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`benchmarks/benchmarks/analysis/distances.py`](https://github.com/MDAnalysis/mdanalysis/blob/aea12a44e3a2f141f95b5d38b3cd1f320b81bdf2/benchmarks/benchmarks/analysis/distances.py):

> [Line 7:1](https://github.com/MDAnalysis/mdanalysis/blob/aea12a44e3a2f141f95b5d38b3cd1f320b81bdf2/benchmarks/benchmarks/analysis/distances.py#L7): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'
> [Line 99:80](https://github.com/MDAnalysis/mdanalysis/blob/aea12a44e3a2f141f95b5d38b3cd1f320b81bdf2/benchmarks/benchmarks/analysis/distances.py#L99): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 101:80](https://github.com/MDAnalysis/mdanalysis/blob/aea12a44e3a2f141f95b5d38b3cd1f320b81bdf2/benchmarks/benchmarks/analysis/distances.py#L101): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 219:80](https://github.com/MDAnalysis/mdanalysis/blob/aea12a44e3a2f141f95b5d38b3cd1f320b81bdf2/benchmarks/benchmarks/analysis/distances.py#L219): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 221:80](https://github.com/MDAnalysis/mdanalysis/blob/aea12a44e3a2f141f95b5d38b3cd1f320b81bdf2/benchmarks/benchmarks/analysis/distances.py#L221): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)



##### Comment last updated at 2021-09-15 07:01:27 UTC

richardjgowers commented:
Hey @Miro-Astore thanks for starting this, it looks great. I'll get it reviewed this weekend

tylerjereddy commented on benchmarks/benchmarks/analysis/distances.py:
```diff
@@ -11,120 +17,243 @@
+    function.
     """
+    u = MDAnalysis.Universe(GRO)
+    tri_clinic_dimensions = u.dimensions
```
What is the purpose of `tri_clinic_dimensions` here and elsewhere? 

Checking this feature branch locally--it seems to be assigned to but never used?

`git grep -E 'tri_clinic_dimensions'`

```
benchmarks/benchmarks/analysis/distances.py:    tri_clinic_dimensions = u.dimensions
benchmarks/benchmarks/analysis/distances.py:    tri_clinic_dimensions = u.dimensions
benchmarks/benchmarks/analysis/distances.py:    tri_clinic_dimensions = u.dimensions
```

Miro-Astore replied:
Should have been deleted, sorry, thanks for catching it.

tylerjereddy commented on benchmarks/benchmarks/analysis/distances.py:
```diff
@@ -11,120 +17,243 @@
+    """Benchmark for the MDAnalysis.analysis.distances.between
+    function.
     """
+    u = MDAnalysis.Universe(GRO)
```
is it necessary to initialize `u = MDAnalysis.Universe(GRO)` both at the class level and inside the `setup()` of each benchmark class?

Miro-Astore replied:
No you're right, sorry these were holdovers from earlier in the writing process and should have been deleted.

tylerjereddy commented on benchmarks/benchmarks/analysis/distances.py:
```diff
@@ -11,120 +17,243 @@
+            self.box_dims = self.u.dimensions
+
+        else:
+            raise Exception("Specify only 'orthogonal', 'triclinic', or None for the box type")
```
Here and elsewhere it seems a bit unusual to raise a generic exception (or any exception, really) in the `setup()` of a benchmark. Since the `asv` benchmarks are not "public" and we control the parameters, plus if the box type is invalid, we should already have error handling for it. `asv` will tend to handle exceptions fairly gracefully, reporting results for benchmarks that do not fail, and filling the tables in with failure entries for those benchmarks that fail.

Miro-Astore replied:
Yes, I thought it might be odd. I'll get rid of them.

tylerjereddy commented on benchmarks/benchmarks/analysis/distances.py:
```diff
@@ -11,120 +17,243 @@
+
+        np.random.seed(17809)
+        self.coords_1 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+        np.random.seed(9008716)
```
Since we can use fairly new NumPy to run benchmarks, we may want to think about using the "newer"/[recommended mode](https://numpy.org/doc/stable/reference/random/new-or-different.html#new-or-different) of random number handling for new benchmarks eventually.

That said, I do find the "old way" far easier to understand, so probably fine for now.

Miro-Astore replied:
Happy to open a new issue if you like. I assume this will be an issue through a lot of the benchmarks module.

tylerjereddy commented:
I should also mention that if timeouts are getting in your way, I'm almost certain `asv` does give us granularity of control on this kind of stuff on a per-benchmark basis, so if you're refactoring just to avoid timeouts you may be able to avoid some setup duplication with a manual setting of the timeouts/repeat threshold parameters.

On the other hand, if you find yourself needing to bump up the timeout limit quite a lot, the benchmarks may be sensible to separate and/or find somewhat smaller representative cases.

Miro-Astore committed:
Fixing up small errors in PBC pull request.

Took out redundant definitions of universe at the class level and
innapropriate raising of Exceptions.

Miro-Astore commented:
I ran the benchmarks with a time out of 10 minutes on my laptop and it still failed, seems like keeping them apart is fine? comments in #2794 indicated that the 1000 atoms was sufficient.

I've just pushed a new commit taking out the redundant definitions of universes and inappropriate raising of exceptions.

richardjgowers commented on benchmarks/benchmarks/analysis/distances.py:
```diff
@@ -11,120 +17,231 @@
         self.ag3 = self.u.atoms[-num_atoms:]
 
-    def time_distance_array(self, num_atoms):
+        np.random.seed(17809)
+        self.coords_1 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+        np.random.seed(9008716)
+        self.coords_2 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+        self.allocated_array_2D = np.empty((num_atoms, num_atoms),
+                                           dtype=np.float64)
+        self.array_shape_1D = int(num_atoms * (num_atoms - 1) / 2.)
+        self.allocated_array_1D = np.empty(self.array_shape_1D,
+                                           dtype=np.float64)
```
If we're generating synthetic coordinates (which isn't a bad idea) then we can probably not actually use an mda.Universe above.

Miro-Astore replied:
Got a quick link for how to easily manufacture dummy universe? A few functions require atom groups.

Miro-Astore replied:
You mentioning this made me play more around with it. The failures in #2794 seems to be due to the fact that we use synthetic coordinates for certain functions causing a memory error.

When we draw samples from `np.random.random_samples` to generate synthetic coordinates we _only_ get numbers between 0 and 1. 

`MDanalysis.analysis.distances.contacts_matrix`  relies on a cutoff distance (default=15). So the benchmark was drawing edges between every atom. The 10 000 particles were causing memory overflow, not a time out issue. 

I can run the benchmark with 30 000 atoms on my 16GB laptop using realistic positions in the data files. The performance of this function seems to be very dependent on data that looks 'real'. 

This will all be in my next commit.

hmacdope commented:
Sorry to chime in here but what do we consider to be a "typical" and "upper limit" problem size? Probably worth a bit of public discussion as this is also a question over at `distopia`.

Miro-Astore commented:
So, I've basically rewritten everything here. I

scal444 commented:
Is this still alive? I have some work on nsgrid I wanted to better quantify in the benchmarks, but I don't want to mess with this file if these improvements are going to go in creating clashes/ duplicating work.

Miro-Astore commented:
Hi, yes I'm still working on it. Sorry for the delays I've been quite busy.
I've basically rewritten the whole thing again. I'll try to push my changes
this weekend. I've made substantial changes to the benchmarks so it might
be best to wait.

Best, Miro

On Thu, Oct 28, 2021 at 10:22 AM Kevin Boyd ***@***.***>
wrote:

> Is this still alive? I have some work on nsgrid I wanted to better
> quantify in the benchmarks, but I don't want to mess with this file if
> these improvements are going to go in creating clashes/ duplicating work.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/3409#issuecomment-953380863>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF6XDG5HOS64KNTEYBNIGFLUJCCURANCNFSM5C6662UA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Miro A. Astore   (he/him)
PhD Candidate | Computational Biophysics
Office 434 A28 School of Physics
University of Sydney

scal444 commented:
> Hi, yes I'm still working on it. Sorry for the delays I've been quite busy. I've basically rewritten the whole thing again. I'll try to push my changes this weekend. I've made substantial changes to the benchmarks so it might be best to wait. Best, Miro
> […](#)
> On Thu, Oct 28, 2021 at 10:22 AM Kevin Boyd ***@***.***> wrote: Is this still alive? I have some work on nsgrid I wanted to better quantify in the benchmarks, but I don't want to mess with this file if these improvements are going to go in creating clashes/ duplicating work. — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub <[#3409 (comment)](https://github.com/MDAnalysis/mdanalysis/pull/3409#issuecomment-953380863)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AF6XDG5HOS64KNTEYBNIGFLUJCCURANCNFSM5C6662UA> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
> -- Miro A. Astore (he/him) PhD Candidate | Computational Biophysics Office 434 A28 School of Physics University of Sydney

Thanks!!! No rush at all, it's just good to know not to try to get something commit-worthy done before rebasing on your changes.
